### PR TITLE
Remote attendance links version 2

### LINKED ISSUE #1281: Add event support for attendance links (and separate into different model)
Also, allow those links to pass the user through to an arbitrary URL transparently (for example, event call links).

Closes #1235 and #1232 by making them irrelevant. Fixes #1392.

### PR DESCRIPTION
**Pull request by JoshOldenburg**
Adds event support for attendance links, allows multiple links per attendable at a time, supports transparent passthroughs, and automatically stores attendance for the person creating the link.

Fixes #1281, #1392, #1232, and #1235.

### CODE REVIEW

**Comment by JoshOldenburg** [comment]
Also fixes #1134.

File: app/Nova/Actions/CreateRemoteAttendanceLink.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+         */
+        return [
+            Text::make('Redirect URL')
+                ->required(false)
+                ->rules('nullable', 'url', 'regex:/^https:\/\/((gatech\.)?bluejeans\.com\/[0-9]+(\/[0-9]+)?|primetime'.
+                    '\.bluejeans\.com\/a2m\/live-event\/[a-z]+|meet\.google\.com\/[-a-z]+)(\?[^@]*)?$/', 'max:1023')
+                ->help('If you put a link to a BlueJeans or Google Meet meeting here, everyone who clicks the '.
+                    'attendance link will be redirected to that meeting after their attendance is recorded. If '.
+                    'you add a redirect URL, do not share that URL directly. Only Google Meet and '.
+                    'BlueJeans calls are supported currently. Contact an administrator for other redirect URLs.'),
```
Comment (kberzinch) [inline app/Nova/Actions/CreateRemoteAttendanceLink.php:100]: Direct to #it-helpdesk

File: app/Policies/RemoteAttendanceLinkPolicy.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    /**
+     * Determine whether the user can view the remote attendance link.
+     */
+    public function view(User $user, RemoteAttendanceLink $attendance): bool
+    {
+        if (! $user->can('read-remote-attendance-links')) {
+            return false;
+        }
+
+        if (is_a($attendance->attendable, 'App\Team') && ! $attendance->attendable->visible) {
```
Comment (kberzinch) [inline app/Policies/RemoteAttendanceLinkPolicy.php:24]:
```suggestion
        if (is_a($attendance->attendable, \App\Team::class) && ! $attendance->attendable->visible) {
```

File: app/Http/Controllers/RemoteAttendanceController.php [hunk 3]
```diff
@@ -41,17 +72,44 @@ public function index(Request $request, string $secret)
...
+            return redirect()->away($link->redirect_url);
+        }
+
         return view(
             'attendance.remote',
             [
-                'message' => 'Your attendance has been recorded for '.$team->name
-                .'. You may now close this page.',
+                'message' => 'Your attendance has been recorded for '.$name.'. You may now close this page. You can '.
+                    'now continue to the meeting:',
```
Comment (kberzinch) [inline app/Http/Controllers/RemoteAttendanceController.php:100]: There should only be one CTA on this page.

**Review by kberzinch — CHANGES_REQUESTED** [review]
Looks good, just a few minor comments

**Review by kberzinch — APPROVED** [review]

File: app/Http/Controllers/RemoteAttendanceController.php [hunk 4]
```diff
@@ -2,33 +2,64 @@
...
+
+        if ($expired) {
+            if ($urlIsValid) {
+                return view(
+                    'attendance.remote',
+                    [
+                        'message' => 'That link is no longer valid so your attendance has not been recorded. Please '.
+                            'ask your project manager for a new one. However, you can still continue to the meeting:',
+                        'linkDestination' => $link->redirect_url,
+                    ]
```
[the remark this answers is not in the document]
Comment (JoshOldenburg) [reply app/Http/Controllers/RemoteAttendanceController.php:48]: Per Slack we're leaving this as is.

**Comment by JoshOldenburg** [comment]
Feedback from user testing:
- Add training and trainers to purpose dropdown
- Clarify "<subteam>" to "<subteam> Subteam" in purpose dropdown
- Hide "Other Purpose" field unless "Other" is selected

File: app/Http/Controllers/RemoteAttendanceController.php [hunk 5]
```diff
@@ -41,17 +72,43 @@ public function index(Request $request, string $secret)
             $att->attendable_type = $attendable_type;
             $att->attendable_id = $attendable_id;
             $att->gtid = $gtid;
-            $att->source = 'secret-link';
+            $att->source = 'secret-link-'.$link->id;
```
Comment (kberzinch) [inline app/Http/Controllers/RemoteAttendanceController.php:75]: Should we maybe put the link ID in a separate column?

File: app/Http/Controllers/RemoteAttendanceController.php [hunk 6]
```diff
@@ -41,17 +72,43 @@ public function index(Request $request, string $secret)
...
             $att->recorded_by = $request->user()->id;
             $att->save();
         }
 
+        $name = is_a($link->attendable, Team::class) || is_a($link->attendable, Event::class) ?
+            $link->attendable->name : 'this event';
```
Comment (kberzinch) [inline app/Http/Controllers/RemoteAttendanceController.php:81]: Why not just use the name always?

File: app/Models/RemoteAttendanceLink.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+     * The attributes that should be mutated to dates.
+     *
+     * @var array<string>
+     */
+    protected $dates = [
+        'created_at',
+        'updated_at',
+        'deleted_at',
+        'expires_at',
+    ];
```
Comment (kberzinch) [inline app/Models/RemoteAttendanceLink.php:48]: Use `$casts`

File: app/Nova/Actions/CreateRemoteAttendanceLink.php [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        $expiration = Carbon::now()->addHours(4);
+
+        // Only on detail, so it will only ever have one model.
+        $attendable = $models->first();
+
+        $link = new RemoteAttendanceLink();
+
+        $link->attendable_type = get_class($attendable);
+        $link->attendable_id = $attendable->id;
+        $link->secret = hash('sha256', random_bytes(64));
```
Comment (kberzinch) [inline app/Nova/Actions/CreateRemoteAttendanceLink.php:48]: I like this approach better ⬇️
https://github.com/RoboJackets/apiary/blob/239405760dd64900156af7c3f79f9c29dd173400/app/Http/Controllers/SquareCheckoutController.php#L213
(I know this was copied from the previous implementation but it's a good opportunity to swap it.)

File: app/Nova/Actions/CreateRemoteAttendanceLink.php [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        // use that value. This deliberately allows empty values.
+        $link->note = $fields->purpose;
+        $link->save();
+        $link->refresh(); // Update id field
+
+        $user = Auth::user();
+        $attExisting = Attendance::where('attendable_type', get_class($attendable))
+            ->where('attendable_id', $attendable->id)
+            ->where('gtid', $user->gtid)
+            ->whereDate('created_at', date('Y-m-d'))->count();
```
Comment (kberzinch) [inline app/Nova/Actions/CreateRemoteAttendanceLink.php:61]: Use `exists`

File: app/Nova/Actions/CreateRemoteAttendanceLink.php [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            $att = new Attendance();
+            $att->attendable_type = get_class($attendable);
+            $att->attendable_id = $attendable->id;
+            $att->gtid = $user->gtid;
+            $att->source = 'secret-link-creation-'.$link->id;
+            $att->recorded_by = $user->id;
+            $att->save();
+        }
+
+        return Action::push('/resources/remote-attendance-links/'.$link->id);
```
Comment (kberzinch) [inline app/Nova/Actions/CreateRemoteAttendanceLink.php:73]: Nice 👍🏻

File: app/Nova/Event.php [hunk 11]
```diff
@@ -88,14 +89,19 @@ public function fields(Request $request): array
...
+
             self::metadataPanel(),
 
             HasMany::make('RSVPs')
                 ->canSee(static function (Request $request): bool {
                     return $request->user()->can('read-rsvps');
                 }),
 
-            HasMany::make('Attendance')
+            MorphMany::make('Attendance')
```
Comment (kberzinch) [inline app/Nova/Event.php:104]: Was this broken before? 🤔

File: app/Nova/RemoteAttendanceLink.php [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+    /**
+     * The list of recommended values for the note field.
+     *
+     * @var array<string>
+     *
+     * @phan-suppress PhanReadOnlyPublicProperty
+     */
+    public static $recommendedNotes = [
+        'Electrical Subteam', 'Mechanical/Mechatronics Subteam', 'Software Subteam', 'Whole Team', 'Training',
+        'Trainers Only', 'Discipline Core (n/a)', 'Event (n/a)',
```
Comment (kberzinch) [inline app/Nova/RemoteAttendanceLink.php:73]: What's with the `(n/a)`?

File: app/Nova/RemoteAttendanceLink.php [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+                ->resolveUsing(static function (string $secret): string {
+                    return route('attendance.remote', ['secret' => $secret]);
+                })
+                ->readonly(static function (Request $request): bool {
+                    return true;
+                }),
+
+            Text::make('Secret')
+                ->onlyOnForms()
+                ->default(hash('sha256', random_bytes(64)))
```
Comment (kberzinch) [inline app/Nova/RemoteAttendanceLink.php:121]: Same as earlier. Should we factor that out to a helper class somewhere?

File: app/Nova/RemoteAttendanceLink.php [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+                    'attendance link will be redirected to that meeting after their attendance is recorded. If '.
+                    'you add a redirect URL, do not share that URL directly. Only Google Meet and '.
+                    'BlueJeans calls are supported currently in the user-facing action.'),
+
+            Text::make('Note')
+                ->hideFromIndex()
+                ->required(false)
+                ->help('This can be used to keep track of what this link was used for more specifically. Press the '.
+                    'down arrow for suggestions.')
+                ->suggestions(self::$recommendedNotes),
```
Comment (kberzinch) [inline app/Nova/RemoteAttendanceLink.php:148]: You don't want to use this on the action?

File: app/Nova/RemoteAttendanceLink.php [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+    protected function metaFields(): array
+    {
+        return [
+            DateTime::make('Created', 'created_at')
+                ->onlyOnDetail(),
+
+            DateTime::make('Last Updated', 'updated_at')
+                ->onlyOnDetail(),
+        ];
+    }
```
Comment (kberzinch) [inline app/Nova/RemoteAttendanceLink.php:168]: These are provided by the parent class

File: app/Nova/RemoteAttendanceLink.php [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+    /**
+     * Get the actions available for the resource.
+     *
+     * @return array<\Laravel\Nova\Actions\Action>
+     */
+    public function actions(Request $request): array
+    {
+        return [];
+    }
```
Comment (kberzinch) [inline app/Nova/RemoteAttendanceLink.php:208]: Delete these

**Review by kberzinch — CHANGES_REQUESTED** [review]
Some quick notes

File: app/Nova/RemoteAttendanceLink.php (hunk 14 above)
Comment (JoshOldenburg) [reply app/Nova/RemoteAttendanceLink.php:148]: No; in fact, I'll switch this to use the select-or-other component as well. I want to strongly encourage using one of the suggestions.

File: app/Nova/Event.php (hunk 11 above)
Comment (JoshOldenburg) [reply app/Nova/Event.php:104]: No; the classes are equivalent. This is more semantically correct.

File: app/Nova/RemoteAttendanceLink.php (hunk 12 above)
Comment (JoshOldenburg) [reply app/Nova/RemoteAttendanceLink.php:73]: I think I originally called the field "Subteam" or something along those lines. I'll remove the `(n/a)`.

File: app/Http/Controllers/RemoteAttendanceController.php (hunk 6 above)
Comment (JoshOldenburg) [reply app/Http/Controllers/RemoteAttendanceController.php]: To keep PHPstan happy

File: app/Http/Controllers/RemoteAttendanceController.php (hunk 5 above)
Comment (JoshOldenburg) [reply app/Http/Controllers/RemoteAttendanceController.php]: Agreed and implemented.

File: app/Nova/RemoteAttendanceLink.php (hunk 13 above)
Comment (JoshOldenburg) [reply app/Nova/RemoteAttendanceLink.php]: Eventually that would probably be good.

File: app/Http/Controllers/RemoteAttendanceController.php [hunk 17]
```diff
@@ -41,17 +72,44 @@ public function index(Request $request, string $secret)
             $att->attendable_type = $attendable_type;
             $att->attendable_id = $attendable_id;
             $att->gtid = $gtid;
-            $att->source = 'secret-link';
+            $att->source = 'remote-attendance-link';
             $att->recorded_by = $request->user()->id;
+            $att->remote_attendance_link_id = $link->id;
```
Comment (kberzinch) [inline app/Http/Controllers/RemoteAttendanceController.php:77]: Define this magic property

File: app/Http/Controllers/RemoteAttendanceController.php [hunk 18]
```diff
@@ -2,33 +2,64 @@
...
             );
         }
 
-        $attendable_type = $team->getMorphClass();
-        $attendable_id = $team->id;
+        $attendable_type = $link->attendable_type;
+        $attendable_id = $link->attendable_id;
         $gtid = $request->user()->gtid;
 
         $attExisting = Attendance::where('attendable_type', $attendable_type)
```
Comment (kberzinch) [inline app/Http/Controllers/RemoteAttendanceController.php:65]: Use `exists`

File: app/Nova/Actions/CreateRemoteAttendanceLink.php [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        // Four hours is longer than most meetings, but not too much longer than that.
+        $expiration = Carbon::now()->addHours(4);
+
+        // Only on detail, so it will only ever have one model.
+        $attendable = $models->first();
+
+        $link = new RemoteAttendanceLink();
+
+        $link->attendable_type = get_class($attendable);
```
Comment (kberzinch) [inline app/Nova/Actions/CreateRemoteAttendanceLink.php:46]:
```suggestion
        $link->attendable_type = $attendable->getMorphClass();
```

File: app/Nova/Actions/CreateRemoteAttendanceLink.php [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+        $link->attendable_id = $attendable->id;
+        $link->secret = bin2hex(openssl_random_pseudo_bytes(32));
+        $link->expires_at = $expiration;
+        $link->redirect_url = $fields->redirect_url;
+        $link->note = $fields->purpose;
+        $link->save();
+        $link->refresh(); // Update id field
+
+        $user = Auth::user();
+        $attExists = Attendance::where('attendable_type', get_class($attendable))
```
Comment (kberzinch) [inline app/Nova/Actions/CreateRemoteAttendanceLink.php:56]:
```suggestion
        $attExists = Attendance::where('attendable_type', $attendable->getMorphClass())
```

File: app/Nova/Actions/CreateRemoteAttendanceLink.php [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+            ->where('attendable_id', $attendable->id)
+            ->where('gtid', $user->gtid)
+            ->whereDate('created_at', date('Y-m-d'))->exists();
+
+        if (! $attExists) {
+            $att = new Attendance();
+            $att->attendable_type = get_class($attendable);
```
Comment (kberzinch) [inline app/Nova/Actions/CreateRemoteAttendanceLink.php:63]:
```suggestion
            $att->attendable_type = $attendable->getMorphClass();
```

File: composer.json [hunk 22]
```diff
@@ -111,6 +112,13 @@
                 "robojackets/*",
                 "oitnetworkservices/*"
             ]
+        },
+        {
+            "type": "vcs",
+            "url": "https://github.com/RoboJackets/select-or-text",
+            "only": [
+                "lynndigital/select-or-custom"
+            ]
```
Comment (kberzinch) [inline composer.json:121]: No longer necessary as discussed in Slack

File: composer.json [hunk 23]
```diff
@@ -25,6 +25,7 @@
         "laravel/slack-notification-channel": "2.3.0",
         "laravel/socialite": "5.1.3",
         "laravel/tinker": "2.5.0",
+        "lynndigital/select-or-custom": "dev-master",
```
Comment (kberzinch) [inline composer.json:28]: They really don't tag this huh 😬

**Review by kberzinch — CHANGES_REQUESTED** [review]
Really close, just some cleanliness notes

File: composer.json (hunk 23 above)
Comment (JoshOldenburg) [reply composer.json:28]: 14 commits and no tags in the whole history of that repo

**Review by kberzinch — APPROVED** [review]

**Pull request merged**